asmcomp/thread_sanitizer.ml

```diff
@@ -0,0 +1,283 @@
+    typ_int
+  | Word_val -> typ_val
+  | Single | Double -> typ_float
+
+let dbg_none = Debuginfo.none
+
+let wrap_entry_exit expr =
+  let call_entry =
+    Cmm_helpers.return_unit dbg_none
+      (Cop
+         ( Cextcall ("__tsan_func_entry", typ_void, [], false),
+           [Creturn_addr],
+           dbg_none ))
+  in
+  let call_exit =
+    Cmm_helpers.return_unit dbg_none
+      (Cop (Cextcall ("__tsan_func_exit", typ_void, [], false), [], dbg_none))
+  in
+  (* [is_tail] is true when the expression is in tail position *)
+  let rec insert_call_exit is_tail = function
+    | Clet (v, e, body) -> Clet (v, e, insert_call_exit is_tail body)
+    | Clet_mut (v, typ, e, body) ->
+      Clet_mut (v, typ, e, insert_call_exit is_tail body)
+    | Cphantom_let (v, e, body) ->
+      Cphantom_let (v, e, insert_call_exit is_tail body)
+    | Cassign (v, body) -> Cassign (v, insert_call_exit is_tail body)
+    | Csequence (op1, op2) -> Csequence (op1, insert_call_exit is_tail op2)
+    | Cifthenelse (cond, t_dbg, t, f_dbg, f, dbg_none) ->
+      Cifthenelse
+        ( cond,
+          t_dbg,
+          insert_call_exit is_tail t,
+          f_dbg,
+          insert_call_exit is_tail f,
+          dbg_none )
+    | Cswitch (e, cases, handlers, dbg_none) ->
+      let handlers =
+        Array.map
+          (fun (handler, handler_dbg) ->
+            insert_call_exit is_tail handler, handler_dbg)
+          handlers
+      in
+      Cswitch (e, cases, handlers, dbg_none)
+    | Ccatch (isrec, handlers, next) ->
+      let handlers =
+        List.map
+          (fun (id, args, e, dbg_none) ->
+            id, args, insert_call_exit is_tail e, dbg_none)
+          handlers
+      in
+      Ccatch (isrec, handlers, insert_call_exit is_tail next)
+    | Cexit (ex, args) ->
+      (* A [Cexit] is like a goto to the beginning of a handler. Therefore, it
+         is never the last thing evaluated in a function; there is no need to
+         insert a call to [__tsan_func_exit] here. *)
+      Cexit (ex, args)
+    | Ctrywith (e, v, handler, dbg_none) ->
+      (* We need to insert a call to [__tsan_func_exit] at the tail of both the
+         body and the handler. If this is a [try ... with] in tail position,
+         then the body expression is not in tail position (as code is inserted
+         at the end of it to pop the exception handler), the handler expression
+         is. *)
+      Ctrywith
+        ( insert_call_exit false e,
+          v,
+          insert_call_exit is_tail handler,
+          dbg_none )
+    | Cop (Capply fn, args, dbg_none) when is_tail ->
+      (* This is a tail call. We insert the call to [__tsan_func_exit] right
+         before the call, but after evaluating the arguments (from right to
+         left). We make an exception for arguments which evaluate to a value of
+         type [Addr], as such values should never be live across a function
+         call or allocation point. *)
```

Comment:
The comment says that arguments of type Addr are treated specially, but I don't see any special treatment below, all arguments are let-bound. Where can we find the code corresponding to this part of the comment?

Reply:
Indeed, we removed this special case when @damiendoligez pointed out that this was unnecessary since TSan instrumentation calls are `[@@unboxed] [@@noalloc]`. Well spotted, thanks! I pushed an update.